File: skills/zoho-blueprint-extractor/lib/gen-yaml.py
#!/usr/bin/env python3
# gen-yaml.py — convertit le JSON brut (sortie de extract-blueprint.js) en YAML normalisé.
# Usage : python3 lib/gen-yaml.py <input.json> <output.yaml>
import json, re, sys

def deHtml(s):
    if not s: return ""
    s = re.sub(r"<[^>]+>", " ", s)
    s = s.replace("&nbsp;", " ").replace("&rsquo;", "’").replace("&amp;", "&")
    return re.sub(r"\s+", " ", s).strip()

def crit(details):
    if not details: return None
    parts = []
    for x in details:
        if isinstance(x, str): parts.append(x.upper())
        elif isinstance(x, dict):
            parts.append(f"{x.get('api_name')} {x.get('comparator')} {x.get('value')}")
    return " ".join(parts)

def q(s): return '"' + str(s).replace('"', "'") + '"'

def main(inp, outp):
    d = json.load(open(inp))
    p = d["process"]; L = []
    L += ["# Blueprint extrait automatiquement (agent-browser + cookies Chrome).",
          "# Source : endpoints internes ProcessFlow.do + FlowTransition.do.", ""]
    L += ["blueprint:",
          f"  name: {q(p.get('Name'))}",
          f"  module: {q(p.get('Module'))}",
          f"  status: {q(p.get('Status'))}",
          f"  continuous: {str(bool(p.get('Continuous'))).lower()}"]
    fld = p.get("Field") or {}
    L.append(f"  state_field: {q(fld.get('DisplayLabel') or fld.get('FieldName') or fld.get('ApiName') or 'Status')}")
    L.append(f"  entry_criteria: {q(deHtml(p.get('entry_criteria')))}")
    L.append(f"  supported_action_types: {p.get('supported_action_types')}")
    L.append("")
    L.append("states:")
    for name in p.get("states", {}).values():
        L.append(f"  - {q(name)}")
    L.append("")
    L.append("transitions:")
    for t in d["transitions"]:
        L.append(f"  - name: {q(t['name'])}")
        L.append(f"    from: {q(t['from'])}")
        L.append(f"    to: {q(t['to'])}")
        L.append(f"    common: {str(bool(t.get('global'))).lower()}")
        L.append(f"    automatic: {str(bool(t.get('auto'))).lower()}")
        owners = ", ".join(o.get("type", "?") for o in (t.get("owners") or []))
        L.append("    before:")
        L.append(f"      owners: [{owners}]")
        if t.get("criteria"):
            L.append(f"      criteria: {q(deHtml(t['criteria']))}")
        L.append("    during:")
        during = t.get("during") or []
        if not during: L.append("      []")
        for f in during:
            if f.get("Type") == "Info":
                L.append(f"      - message: {q(deHtml(f.get('Info'))[:300])}")
            elif f.get("Type") == "Field":
                c = crit(f.get("CriteriaDetails"))
                L.append(f"      - field_validation: {q(c)}" if c else "      - field: (requis)")
        L.append("    after:")
        A = t.get("actions") or {}
        nz = False
        for k, v in A.items():
            if isinstance(v, list) and v:
                nz = True
                for item in v:
                    label = item.get("Name") or item.get("name") or item.get("url") or k
                    L.append(f"      - type: {k}")
                    L.append(f"        detail: {q(str(label)[:140])}")
                    for pk in ("url", "method", "dueDate", "remindBefore", "priority",
                               "status", "func_name", "scriptName", "scriptId"):
                        if pk in item: L.append(f"        {pk}: {q(item[pk])}")
        if not nz: L.append("      []")
        L.append("")
    open(outp, "w").write("\n".join(L))
    print(f"YAML écrit : {outp}  ({len(d['transitions'])} transitions)")

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python3 lib/gen-yaml.py <input.json> <output.yaml>"); sys.exit(1)
    main(sys.argv[1], sys.argv[2])

File: skills/zoho-blueprint-extractor/lib/gen-yaml.mjs
#!/usr/bin/env node
// gen-yaml.mjs — convertit le JSON brut (sortie de extract-blueprint.js) en YAML normalisé.
// Node only (pas de dépendance Python). Usage : node lib/gen-yaml.mjs <input.json> <output.yaml>
import { readFileSync, writeFileSync } from "node:fs";

const deHtml = (s) =>
  !s ? "" :
  s.replace(/<[^>]+>/g, " ").replace(/&nbsp;/g, " ").replace(/&rsquo;/g, "’")
   .replace(/&amp;/g, "&").replace(/\s+/g, " ").trim();

const crit = (details) => {
  if (!details) return null;
  return details.map((x) =>
    typeof x === "string" ? x.toUpperCase()
    : `${x.api_name} ${x.comparator} ${x.value}`).join(" ");
};
const q = (s) => '"' + String(s ?? "").replace(/"/g, "'") + '"';

const [, , inp, outp] = process.argv;
if (!inp || !outp) { console.error("Usage: node lib/gen-yaml.mjs <input.json> <output.yaml>"); process.exit(1); }

const d = JSON.parse(readFileSync(inp, "utf-8"));
const p = d.process;
const L = [];
L.push("# Blueprint extrait automatiquement (agent-browser + cookies / session navigateur).");
L.push("# Source : endpoints internes ProcessFlow.do + FlowTransition.do.", "");
L.push("blueprint:");
L.push(`  name: ${q(p.Name)}`);
L.push(`  module: ${q(p.Module)}`);
L.push(`  status: ${q(p.Status)}`);
L.push(`  continuous: ${!!p.Continuous}`);
const fld = p.Field || {};
L.push(`  state_field: ${q(fld.DisplayLabel || fld.FieldName || fld.ApiName || "Status")}`);
L.push(`  entry_criteria: ${q(deHtml(p.entry_criteria))}`);
L.push(`  supported_action_types: ${JSON.stringify(p.supported_action_types || [])}`);
L.push("");
L.push("states:");
for (const name of Object.values(p.states || {})) L.push(`  - ${q(name)}`);
L.push("");
L.push("transitions:");
for (const t of d.transitions) {
  L.push(`  - name: ${q(t.name)}`);
  L.push(`    from: ${q(t.from)}`);
  L.push(`    to: ${q(t.to)}`);
  L.push(`    common: ${!!t.global}`);
  L.push(`    automatic: ${!!t.auto}`);
  const owners = (t.owners || []).map((o) => o.type || "?").join(", ");
  L.push("    before:");
  L.push(`      owners: [${owners}]`);
  if (t.criteria) L.push(`      criteria: ${q(deHtml(t.criteria))}`);
  L.push("    during:");
  const during = t.during || [];
  if (!during.length) L.push("      []");
  for (const f of during) {
    if (f.Type === "Info") L.push(`      - message: ${q(deHtml(f.Info).slice(0, 300))}`);
    else if (f.Type === "Field") {
      const c = crit(f.CriteriaDetails);
      L.push(c ? `      - field_validation: ${q(c)}` : "      - field: (requis)");
    }
  }
  L.push("    after:");
  const A = t.actions || {};
  let nz = false;
  for (const [k, v] of Object.entries(A)) {
    if (Array.isArray(v) && v.length) {
      nz = true;
      for (const item of v) {
        const label = item.Name || item.name || item.url || k;
        L.push(`      - type: ${k}`);
        L.push(`        detail: ${q(String(label).slice(0, 140))}`);
        for (const pk of ["url", "method", "dueDate", "remindBefore", "priority",
                          "status", "func_name", "scriptName", "scriptId"])
          if (pk in item) L.push(`        ${pk}: ${q(item[pk])}`);
      }
    }
  }
  if (!nz) L.push("      []");
  L.push("");
}
writeFileSync(outp, L.join("\n"));
console.log(`YAML écrit : ${outp}  (${d.transitions.length} transitions)`);
